### DISCUSSION TITLE
How to set auxclasspath in spotbugs gradle plugin?

### DISCUSSION THREAD

**Discussion by LiuGavin** [spotbugs/spotbugs-gradle-plugin#677] [General]
currently when I run spotbugs on my project, it will report lots of classes names printed following "The following classes needed for analysis were missing:".   After digging around, I find that I have to manually overwrite the ./build/spotbugs/auxclasspath/spotbugsMain file with jar files under my gradle cache directory -- I don't know why spotbugs will look for the jar files there either. But anyway, if I manually put the jar files list in this spotbugsMain file then  spotbugs run without reporting any warning messages any more. But then I realized that this file would be overwritten each time the gradle run. So my question is:
1. how to disable spotbugs to look for the gradle cache directory?
2. If 1 is not feasible, how to specify the auxclasspath in gradle config file without manual changes?

Thanks

**Comment by KengoTODA** [comment]
> how to disable spotbugs to look for the gradle cache directory?

You do not have to disable it. It is expected behavior.

By default, `auxClassPaths` is configured based on the compile-scope dependencies of the target sourceSet. And Gradle put jar files of dependencies into the cache file. So SpotBugs will use jar files in the cache dir.

> If 1 is not feasible, how to specify the auxclasspath in gradle config file without manual changes?

Just like as follows:

```
tasks.spotbugsMain {
  auxClassPaths = sourceSets.main.compileClasspath
}
```

We use compile classpath by default, refer to #115 to learn about the motivation.

**Reply by LiuGavin** [reply]
I added your suggested auxClassPaths setting above but still the same...
```
> Task :spotbugsMain
The following classes needed for analysis were missing:
  org.springframework.boot.web.servlet.support.SpringBootServletInitializer
  org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
  org.springframework.web.servlet.HandlerInterceptor
  org.springframework.web.servlet.handler.HandlerInterceptorAdapter
.....
  com.google.common.base.Predicate
  com.google.common.base.Predicates
....
....
```
I just want spotBugs anaylyze my classes, below is my spotbuts settings-->
```
spotbugs( ) {
    toolVersion = "4.5.3"
    ignoreFailures = true
    showStackTraces = false
    showProgress = false
    effort = 'default'
    reportLevel = 'low'
    excludeFilter = file("${rootProject.projectDir}/findbugs_exclude_filter.xml")
    onlyAnalyze = [ 'com.myproject.*' ]
    jvmArgs = [ '-Duser.language=en' ]
}
spotbugsMain {
    dependsOn compileJava, compileTestJava
    classes = files("$rootProject.projectDir/build/classes/java/main", "$rootProject.projectDir/sub-project/build/classes/java/main")
    reports {
        html {
            required = true
            outputLocation = file("$rootProject.projectDir/build/reports/spotbugs/main/spotbugs.html")
            stylesheet = 'fancy-hist.xsl'
        }
    }
}

tasks.spotbugsMain{
    auxClassPaths = sourceSets.main.compileClasspath
}
```

**Reply by KengoTODA** [reply]
Could you provide a [mcve project](https://stackoverflow.com/help/minimal-reproducible-example) to reproduce the issue?

**Reply by LiuGavin** [reply]
I tried to create a mcve project but failed to reproduce the issue. but at last, I find out that if I merge my sub-module into the main one(so now I only have one module), it works. But in my mcve project the 'sub-module layout' also work. There must be some tasks defined in my sub-module that conflicting with the spotbugs.
